src/rust/engine/process_execution/src/local.rs

```diff
@@ -62,6 +63,7 @@ impl CommandRunner {
       named_caches,
       cleanup_local_dirs,
       platform: Platform::current().unwrap(),
+      spawn_lock: Arc::new(RwLock::new(true)),
```

Comment:
It looks like this lock should only ever be acquired for a few milliseconds, but if we think it might be any longer than that (slow disks?) should probably use https://docs.rs/tokio/0.2.20/tokio/sync/struct.RwLock.html here to avoid blocking the tokio thread.

Reply:
OK. I was going by this from the horses mouth: https://github.com/tokio-rs/mini-redis/blob/master/src/db.rs#L27-L41
That said, this may not be a "**usually**", since, as you say, it's not inconceivable to have a binary for exec to load that takes a while. I'm particularaly thinking of the ML massive python pex case.

Reply:
> That said, this may not be a "usually", since, as you say, it's not inconceivable to have a binary for exec to load that takes a while. I'm particularaly thinking of the ML massive python pex case.

Hopefully the actual paging in of the whole executable would happen after the spawn had completed, but clearly it needs to have at least stat'd the file to decide whether to succeed or fail. So yea.

Reply:
Actually - wait a sec. Tokio themselves designed spawn as synchronous function hat returns a future (Child). Our lock wraps exactly that duration of synchrony and nothing else. So if they are right to do this, we are right to wrap it in a synchronous lock, right?

Reply:
I think the fear for a regular mutex is that if we end up blocking on a bunch of readers (or blocking readers on a writer), we end up clogging up the runtime thread while we wait? So for synchronisation purposes, a synchronous lock is fine, but it will cause problems for other tasks on the runtime.

Reply:
Gotcha - that makes sense.

Reply:
Yea. Between the lines of https://github.com/tokio-rs/mini-redis/blob/master/src/db.rs#L27-L41 is that something like a syscall "should" be made asynchronously. And so saying that "nothing asynchronous occurs under the lock" is akin to saying that nothing "slow enough to be worth making asynchronous" happens under the lock.

Reply:
Fixed.